### DISCUSSION TITLE
Can I run multiple APs?

### DISCUSSION THREAD

**Discussion by fadyanwar** [micropython/micropython#12829] [Using MicroPython]
Hello,
I was experimenting with Raspi Pico and was trying to run multiple wifi APs simultaneously using the below test code, but only got one AP visible. So was wondering if it's supported. Thank you!

import network
import machine

ssid = 'MicroPython-AP'
password = '123456789'

ap = network.WLAN(network.AP_IF)
ap.config(essid=ssid, password=password)
ap.active(True)

ssid2 = 'MicroPython-AP2'

ap2 = network.WLAN(network.AP_IF)
ap2.config(essid=ssid2, password=password)
ap2.active(True)

ssid3 = 'MicroPython-AP3'

ap3 = network.WLAN(network.AP_IF)
ap3.config(essid=ssid3, password=password)
ap3.active(True)

**Comment by jimmo** [comment]
No, there is only a single access point (and a single station). This is a hardware limitation.

The `network.WLAN` constructor actually returns singleton instances... i.e. you will see if you run

```
>>> ap == ap2
True
```

**Reply by fadyanwar** [reply]
Thanks for quick answer. I was wondering if it's really a hardware limitation, as I've seen other projects using the same ESP8266 module but still can run multiple APs. See this one for instance. https://github.com/SpacehuhnTech/esp8266_deauther

**Reply by jimmo** [reply]
The Pico W doesn't use an esp8266.

Also I don't think the esp8266 supports multiple concurrent APs either... there's a single API `wifi_softap_set_config` which takes a structure that only allows configuration of one API. I suspect most of Deauther's magic happens on the station interface (i.e. sending the deauth payloads) and the documentation notes that you can lose connection while running an attack.

**Reply by fadyanwar** [reply]
Is there is  way to achieve the same in MicroPython?

**Reply by jimmo** [reply]
Yes you can change the configuration of the AP.

```py
ap = network.WLAN(network.AP_IF)
# Initial config.
ap.config(essid=ssid, password=password)
ap.active(True)
...
# Update the AP config.
ap.active(False)
ap.config(essid=ssid2, password=password)
ap.active(True)
```

**Reply by fadyanwar** [reply]
Just tested it using the below code

```

import network
import machine

ssid = 'hello'
password = '123456789'

ap = network.WLAN(network.AP_IF)
ap.config(essid=ssid, password=password)
ap.active(True)

ssid2 = 'world'

ap.active(False)
ap.config(essid=ssid2, password=password)
ap.active(True)

```

Only "world" is visible

**Reply by jimmo** [reply]
@fadyanwar That's right. But "hello" would have been visible before you changed it to "world".

**Reply by fadyanwar** [reply]
I would imagine so, but how would this be considered concurrent?

**Reply by jimmo** [reply]
It isn't.

Neither the esp8266 nor the cyw43 in the pico w support concurrent APs. I explained above how Deauther could (non-concurrently!) act as different APs, and you asked how to achieve that in MicroPython, and this is how you do it.
